### PR TITLE
[rfr] Project quotas

### PR DESCRIPTION
**Pull request by haphan**

### CODE REVIEW

**Comment by haphan** [comment]
@jamiehannaford Appreciate if you have time to review this PR.

File: src/Networking/v2/Models/Quota.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+namespace OpenStack\Networking\v2\Models;
+
+use OpenStack\Common\Resource\Deletable;
+use OpenStack\Common\Resource\OperatorResource;
+use OpenStack\Common\Resource\Retrievable;
+use OpenStack\Common\Resource\Updateable;
+use OpenStack\Networking\v2\Api;
+
+/**
+ * Represents a Nova v2 Quota
```
Comment (ernesto-pama) [inline src/Networking/v2/Models/Quota.php]: Please change Nova to Neutron

File: src/Compute/v2/Api.php [hunk 2]
```diff
@@ -668,4 +668,66 @@ public function getHypervisor(): array
             'params' => ['id' => $this->params->urlId('hypervisor')]
         ];
     }
+
+    public function getQuotaSet()
```
Comment (jamiehannaford) [inline src/Compute/v2/Api.php]: needs return types

File: src/Compute/v2/Api.php [hunk 3]
```diff
@@ -668,4 +668,66 @@ public function getHypervisor(): array
...
+                'tenantId' => $this->params->urlId('quota-sets')
+            ]
+        ];
+    }
+
+    public function getQuotaSetDetail()
+    {
+        return [
+            'method'  => 'GET',
+            'path'    => 'os-quota-sets/{tenantId}/detail',
```
Comment (jamiehannaford) [inline src/Compute/v2/Api.php]: this is fine, but if you have large operation data structs in the future, it's better to do:

```php
public function getQuotaSetDetail()
{
   $data = self::getQuotaSet();
   $data['url'] += '/detail';
   return $data;
}
```

File: src/Compute/v2/Models/QuotaSet.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ * Represents a Compute v2 Quota Set
+ *
+ * @property \OpenStack\Compute\v2\Api $api
+ */
+class QuotaSet extends OperatorResource implements Retrievable, Updateable, Deletable
+{
+    /**
+     * The number of allowed instance cores for each tenant.
+     *
+     * @var  int|array
```
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: how can it be both an int and array? then again...openstack 😛

File: src/Compute/v2/Models/QuotaSet.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    public $cores;
+
+    /**
+     * The number of allowed fixed IP addresses for each tenant.
+     * Must be equal to or greater than the number of allowed instances.
+     *
+     * @deprecated Since 2.35
+     * @var  int|object
```
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: same here

File: src/Compute/v2/Models/QuotaSet.php [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    public $fixedIps;
+
+    /**
+     * The number of allowed floating IP addresses for each tenant.
+     *
+     * @deprecated Since 2.35
```
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: what does this mean?

File: src/Compute/v2/Models/QuotaSet.php [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        'floating_ips'                => 'floatingIps',
+        'injected_file_content_bytes' => 'injectedFileContentBytes',
+        'injected_file_path_bytes'    => 'injectedFilePathBytes',
+        'injected_files'              => 'injectedFiles',
+        'key_pairs'                   => 'keyPairs',
+        'metadata_items'              => 'metadataItems',
+        'security_group_rules'        => 'securityGroupRules',
+        'security_groups'             => 'securityGroups',
+        'server_group_members'        => 'serverGroupMembers',
+        'server_groups'               => 'serverGroups',
```
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: we should really add some kind of conversion logic in Common that handles snake_case -> camelCase

File: src/Networking/v2/Models/Quota.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+     * @var string
+     */
+    public $tenantId;
+
+    /**
+     * @var int
+     */
+    public $subnetpool;
+
+
```
Comment (jamiehannaford) [inline src/Networking/v2/Models/Quota.php]: no need for extra line

File: src/Compute/v2/Api.php (hunk 2 above)
Comment (haphan) [inline src/Compute/v2/Api.php]: 👍

File: src/Compute/v2/Api.php (hunk 3 above)
Comment (haphan) [inline src/Compute/v2/Api.php]: 👍

File: src/Compute/v2/Models/QuotaSet.php (hunk 4 above)
Comment (haphan) [inline src/Compute/v2/Models/QuotaSet.php]: Exactly, but then this is how Openstack implemented it. 

For `os-quota-sets/{tenantId}` returns `int`; `os-quota-sets/{tenantId}/detail` returns an array.  See https://developer.openstack.org/api-ref/compute/?expanded=show-the-detail-of-quota-detail#show-the-detail-of-quota

File: src/Compute/v2/Models/QuotaSet.php (hunk 6 above)
Comment (haphan) [inline src/Compute/v2/Models/QuotaSet.php]: See also https://developer.openstack.org/api-ref/compute/?expanded=show-the-detail-of-quota-detail#show-the-detail-of-quota

I think the logic here is that network related quotas such as `floating_ips` and `fixed_ips` shall be managed by Networking service;

I put `@deprecated` flag here to discourage people in using these attributes.

File: src/Networking/v2/Models/Quota.php (hunk 8 above)
Comment (haphan) [inline src/Networking/v2/Models/Quota.php]: 👍

File: src/Compute/v2/Models/QuotaSet.php (hunk 4 above)
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: Hmm, that's a crappy UX for consumers of that API. 

How about we introduce a `Quota` model that encapsulates `in_use`, `limit` and `reserved` fields? If the API returns just an int (e.g. `os-quota-sets/{tenantId}`), that populates just the `limit` field?

File: src/Compute/v2/Models/QuotaSet.php (hunk 6 above)
Comment (jamiehannaford) [inline src/Compute/v2/Models/QuotaSet.php]: Ah I see. To clear up confusion, can we expand the comment to:

```
@deprecated Since Nova v2.35. This attribute will eventually move to Neutron, it is advised you do not use this.
```

File: src/Compute/v2/Models/QuotaSet.php (hunk 4 above)
Comment (haphan) [inline src/Compute/v2/Models/QuotaSet.php]: We certainly can implement that way, but then at the same time I don't want the SDK to diverse from the API response.

File: src/Compute/v2/Models/QuotaSet.php (hunk 6 above)
Comment (haphan) [inline src/Compute/v2/Models/QuotaSet.php]: 👍

File: src/Compute/v2/Models/QuotaSet.php (hunk 7 above)
Comment (haphan) [inline src/Compute/v2/Models/QuotaSet.php]: Agree. This issue will be addressed in another PR.

**Comment by haphan** [comment]
@jamiehannaford All good?

**Comment by jamiehannaford** [comment]
@haphan Sure, is your username on there `haphan` too?
